007-OFEP-flag-change-events.md

```diff
@@ -0,0 +1,74 @@
+## 007-OFEP-flag-change-events
+
+## State: DRAFTING
+
+Some flag SDKs support listening for flag value changes or general configuration changes ([launchdarkly](https://docs.launchdarkly.com/sdk/features/flag-changes), [cloudbees](https://docs.cloudbees.com/docs/cloudbees-feature-management/latest/reporting/configuration-fetched-handler), flagd). This can allow us to use an event-based paradigm for consuming flags. Client apps may use feature flags for characteristics that aren't specifically tied to a user-action, making imperative flag evaluation a less-than-ideal solution. A server application could listen to a flag that changes some operational behavior.
+
+Examples:
+
+- A web client's banner color could be updated when an associated flag is changed, without any user action or a page reload.
+- A service could subscribe to a flag that controls its global log-level.
+- A subsystem could be restarted with the new flag value.
+
+## Design
+
+The provider interface and the OpenFeature client would be extended to have new functionality to register handlers for a set of events defined by the SDK (`ProviderEvents`). When a _application author_ registers a handler on a client, the client the client maintains this handler. The provider emits events or runs a callback indicating that it received a certain event, optionally providing data associated with that event. The callbacks registered with the client are then invoked with this data (`EventData`).
+
+In the case of the aforementioned flag systems, flags are not actually evaluated when configurations are changed. One reason is that no dynamic context can be reasonably provided in the case of events, since the event is driven by a change in the flag management system, not a user-action. In these cases, the `EventData` would not contain flag values and the application author would have to evaluate flags in the registered handler. This is consistent with the event APIs already existing in those systems.
```

Comment:
Counterpoint: in the mobile world, we don't evaluate on the client. "this thing changed" *is* accompanied with values, because we only dispatch on the things that we know about the user (device type and app version).

Reply:
This is a good point. I think that to maintain as similar an API as possible between client and server impls, it's ideal not to include flag values in the event. In the case of mobile web, as you point out, the flags are probably cached. As long as we consider race conditions, and define our ordering rigorously, I'd hope the cache of flags would be up to date, and subsequent evaluations after the events fire would return accurate flag values.